// inotify/win32/index.js
var fs = require('fs'),
	path = require('path');
var log = require('util/logger').log;
var CONST = require('../CONST');

var inotify = function(_path,fn_event){
	var files = {};
	var tt,delay = 10;
	this.fsWatcher = fs.watch(_path, function (event, name) {
		console.log(event,name);
		clearTimeout(tt);
		var type = null;
		tt = setTimeout(function (){
			console.log(1,event,name);return;
			if(name != null){
				var fullname = path.join(_path, name);
				if(fs.existsSync(fullname)){
					var isDir = fs.statSync(fullname).isDirectory();
					if(event == 'rename'){
						type = CONST.EVENT_CREATE;
					}else if(event == 'change'){
						type = CONST.EVENT_CHANGE;
					}
					console.log(2,type,isDir);
					if(type){
						files[name] = isDir;
						console.log(3,files);
						fn_event(type,{
			            	fullname: fullname,
			            	filename: name,
			            	isDir: isDir
						});
						return;
					}
				}
			}

			var deleteFileName,isDir = false;
			if(fullname){
				deleteFileName = name;
				isDir = files[name];
			}else{
				var newFiles = fs.readdirSync(_path);
				for(var i in files){
					if(newFiles.indexOf(i) == -1){
						deleteFileName = i;
						isDir = files[i];
						break;
					}
				}
			}
			
			if(deleteFileName != null){
				delete files[deleteFileName];
				
				fn_event(CONST.EVENT_REMOVE,{
	            	fullname: path.join(_path,deleteFileName),
	            	filename: deleteFileName,
	            	isDir: isDir
				});
			}
		},delay);
	});
	fs.readdir(_path,function(err,_files){
		if(!err){
			_files.forEach(function(file){
				var stat = fs.statSync(path.join(_path,file));
				files[file] = stat.isDirectory();
			});
		}
	});
}
inotify.prototype.remove = function(){
}

exports.inotify = inotify;
